### PR TITLE
get clcache to work on AppVeyor Windows

### PR DESCRIPTION
**Pull request by Be-ing**

### CODE REVIEW

**Comment by Be-ing** [comment]
By the way, caching the build environment does not work because it is over the cache size limit of 1 GB.

**Comment by Be-ing** [comment]
Somehow this still is not working...
```
[00:01:38] cd %APPVEYOR_BUILD_FOLDER%
[00:01:39] clcache -M 4294967296
[00:01:39] clcache -s
[00:01:39] clcache statistics:
[00:01:39]   current cache dir         : Disk cache at C:\Users\appveyor\clcache
[00:01:39]   cache size                : 94,224,983 bytes
[00:01:39]   maximum cache size        : 4,294,967,296 bytes
[00:01:39]   cache entries             : 19
[00:01:39]   cache hits                : 0
[00:01:39]   cache misses
[00:01:39]     total                      : 33134
[00:01:39]     evicted                    : 0
[00:01:39]     header changed             : 393
[00:01:39]     source changed             : 32741
[00:01:39]   passed to real compiler
[00:01:39]     called w/ invalid argument : 0
[00:01:39]     called for preprocessing   : 82
[00:01:39]     called for linking         : 845
[00:01:39]     called for external debug  : 287
[00:01:39]     called w/o source          : 0
[00:01:39]     called w/ multiple sources : 0
[00:01:39]     called w/ PCH              : 0
[00:01:39] set CLCACHE_SERVER=1
[00:01:39] set CLCACHE_HARDLINK=1
```
CMake successfully detects clcache:
```
[00:01:43] -- Found clcache: C:/Python37-x64/Scripts/clcache.exe
[00:01:43] -- Support for clcache: ON
...
[00:01:52] CLCACHE_EXECUTABLE:FILEPATH=C:/Python37-x64/Scripts/clcache.exe
[00:01:52] CLCACHE_SUPPORT:BOOL=ON
```

@JosepMaJAZ do you have any ideas? Does clcache work locally for you?

**Comment by Be-ing** [comment]
I don't know why this isn't working. Maybe we can use [cclash](https://github.com/inorton/cclash) instead?

**Comment by Be-ing** [comment]
clcache [documentation](https://github.com/frerich/clcache#how-clcache-works) says:
```quote
    The /link switch must not be present
    The /c switch must be present
    The /Zi switch must not be present (/Z7 is okay though)
```
I do not see the `/c` switch in the compilation commands. Any ideas how to make this happen?

**Comment by Holzhaus** [comment]
Add it to target_compile_options in Cmakelists.txt.

**Comment by Be-ing** [comment]
Still not working... I'm at a loss now.

**Comment by Be-ing** [comment]
There were some cache hits finally:
```
clcache -s
clcache statistics:
  current cache dir         : Disk cache at C:\Users\appveyor\clcache
  cache size                : 4,057,556,157 bytes
  maximum cache size        : 4,294,967,296 bytes
  cache entries             : 618
  cache hits                : 241
  cache misses
    total                      : 45771
    evicted                    : 0
    header changed             : 1934
    source changed             : 43837
  passed to real compiler
    called w/ invalid argument : 0
    called for preprocessing   : 114
    called for linking         : 1176
    called for external debug  : 399
    called w/o source          : 0
    called w/ multiple sources : 0
    called w/ PCH              : 0
```
but the build still took 45 minutes...

**Comment by Holzhaus** [comment]
Spawning processes is very slow on windows, you should re-enable the server mode.

**Comment by Be-ing** [comment]
I manually reran the same build and there were only 10 cache more cache hits? What is going on??
```
clcache -s
clcache statistics:
  current cache dir         : Disk cache at C:\Users\appveyor\clcache
  cache size                : 4,100,598,700 bytes
  maximum cache size        : 4,294,967,296 bytes
  cache entries             : 619
  cache hits                : 251
  cache misses
    total                      : 46553
    evicted                    : 0
    header changed             : 2715
    source changed             : 43838
  passed to real compiler
    called w/ invalid argument : 0
    called for preprocessing   : 116
    called for linking         : 1197
    called for external debug  : 406
    called w/o source          : 0
    called w/ multiple sources : 0
    called w/ PCH              : 0
Collecting artifacts...
```

> Try reenabling the server mode: https://github.com/frerich/clcache/wiki/Performance#use-clcachesrv

Yes I will try that next.

**Comment by Be-ing** [comment]
After reenabling server mode and hardlinking, more cache hits but still the build took 44 minutes...
```
-results.xml))
clcache -s
clcache statistics:
  current cache dir         : Disk cache at C:\Users\appveyor\clcache
  cache size                : 3,911,602,140 bytes
  maximum cache size        : 4,294,967,296 bytes
  cache entries             : 595
  cache hits                : 298
  cache misses
    total                      : 47298
    evicted                    : 0
    header changed             : 3459
    source changed             : 43839
  passed to real compiler
    called w/ invalid argument : 0
    called for preprocessing   : 118
    called for linking         : 1218
    called for external debug  : 413
    called w/o source          : 0
    called w/ multiple sources : 0
    called w/ PCH              : 0
Collecting artifacts...
```

**Comment by Holzhaus** [comment]
I wonder why there are so many cache misses... You didn't change the code, in between there should be a lot more cache hits.

**Comment by Be-ing** [comment]
Repeat build took 34 minutes, which is slightly faster...
```
clcache -s
clcache statistics:
  current cache dir         : Disk cache at C:\Users\appveyor\clcache
  cache size                : 3,977,235,524 bytes
  maximum cache size        : 4,294,967,296 bytes
  cache entries             : 592
  cache hits                : 430
  cache misses
    total                      : 47958
    evicted                    : 0
    header changed             : 4117
    source changed             : 43841
  passed to real compiler
    called w/ invalid argument : 0
    called for preprocessing   : 120
    called for linking         : 1239
    called for external debug  : 420
    called w/o source          : 0
    called w/ multiple sources : 0
    called w/ PCH              : 0
```
I am puzzled why there are so many cache misses. I don't know what else to try. Shall we go ahead and merge this?

**Comment by Be-ing** [comment]
Even though this is only a slight improvement, I think we should go ahead and merge it now. I am out of ideas to improve it further. Hopefully someone else can pick it up in the future.

**Comment by poelzi** [comment]
Could it be, that it writes a header file with compile time or something that will change the preprocessed output?

**Comment by Be-ing** [comment]
That should only affect a few files.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1000,6 +1000,8 @@ if(WIN32)
   target_link_libraries(mixxx-lib PUBLIC shell32)
 
   if(MSVC)
+    # needed for clcache
+    target_compile_options(mixxx-lib PUBLIC /c)
```
Comment (Holzhaus) [inline CMakeLists.txt:1004]: No idea what `/c` does? Does it have any bad side effects? Should we only add it if `CLCACHE_SUPPORT` is `ON`?

File: appveyor.yml [hunk 2]
```diff
@@ -252,7 +255,7 @@ for:
       -DSTATIC_DEPS=ON
       -G "Ninja"
       ..
-    - cmake --build .
+    - cmake --build . -v
```
Comment (Holzhaus) [inline appveyor.yml:258]: Remove the `-v` please

**Comment by Holzhaus** [comment]
Maybe we should switch to cclash in the long run, as clcache seem so be abandoned (repo is archived). But no need to do that in this PR.

**Comment by Holzhaus** [comment]
@Be-ing Ready to merge, even if the cache miss count is high?

File: appveyor.yml (hunk 2 above)
Comment (Holzhaus) [reply appveyor.yml:258]: Ping

Comment (Holzhaus) [reply appveyor.yml:258]: Ping

**Pull request merged**